File: sdks/community/fraiseql-kotlin/src/test/kotlin/com/fraiseql/SchemaRoundtripTest.kt
package com.fraiseql

import com.fraiseql.schema.Schema
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

/**
 * SDK-3: Schema roundtrip golden test.
 *
 * Exercises the full decorator → JSON export pipeline: register a type
 * with fields (including a scoped field), export to JSON, and verify that
 * the output matches the expected schema.json structure exactly.
 *
 * This is the contract between the SDK and the fraiseql-cli compiler.
 * If the SDK produces malformed JSON the compiler rejects it — but
 * without this test that failure is silent during SDK development.
 */
class SchemaRoundtripTest {

    @BeforeEach
    fun setUp() {
        Schema.reset()
    }

    @AfterEach
    fun tearDown() {
        Schema.reset()
    }

    @Test
    fun `full decorator to export pipeline produces expected schema json structure`() {
        // Register a realistic type with a mix of field types, including a scoped field.
        Schema.registerType("Article", mapOf(
            "id"    to mapOf("type" to "ID",     "nullable" to false),
            "title" to mapOf("type" to "String", "nullable" to false),
            "body"  to mapOf("type" to "String", "nullable" to true),
            "email" to mapOf("type" to "String", "nullable" to false, "scope" to "read:Article.email"),
        ), "A published article")

        val jsonStr = Schema.exportTypes(pretty = true)
        val parsed = Json.parseToJsonElement(jsonStr).jsonObject

        // Must contain exactly the `types` key — no compiler-reserved keys
        assertTrue(parsed.containsKey("types"),    "output must have `types` key")
        assertFalse(parsed.containsKey("queries"),    "output must NOT have `queries`")
        assertFalse(parsed.containsKey("mutations"),  "output must NOT have `mutations`")
        assertFalse(parsed.containsKey("observers"),  "output must NOT have `observers`")
        assertFalse(parsed.containsKey("security"),   "output must NOT have `security`")
        assertFalse(parsed.containsKey("federation"), "output must NOT have `federation`")

        // Exactly one type was registered
        val types = parsed["types"]!!.jsonArray
        assertEquals(1, types.size, "exactly one type was registered")

        // Verify Article type structure
        val article = types[0].jsonObject
        assertEquals("Article",            article["name"]!!.jsonPrimitive.content, "type name must be Article")
        assertEquals("A published article", article["description"]!!.jsonPrimitive.content, "description must round-trip")

        // All four fields must be present
        val fields     = article["fields"]!!.jsonArray
        val fieldNames = fields.map { it.jsonObject["name"]!!.jsonPrimitive.content }
        assertTrue(fieldNames.contains("id"),    "field `id` must be present")
        assertTrue(fieldNames.contains("title"), "field `title` must be present")
        assertTrue(fieldNames.contains("body"),  "field `body` must be present")
        assertTrue(fieldNames.contains("email"), "field `email` must be present")

        // The scoped field must carry its scope annotation
        val emailField = fields.map { it.jsonObject }.firstOrNull {
            it["name"]!!.jsonPrimitive.content == "email"
        }
        assertNotNull(emailField, "email field must be present in output")
        assertEquals("read:Article.email", emailField!!["scope"]!!.jsonPrimitive.content,
            "scope annotation must round-trip")
    }

    @Test
    fun `multiple registered types all appear with correct names`() {
        Schema.registerType("User", mapOf(
            "id"   to mapOf("type" to "ID",     "nullable" to false),
            "name" to mapOf("type" to "String", "nullable" to false),
        ), "System user")

        Schema.registerType("Post", mapOf(
            "id"    to mapOf("type" to "ID",     "nullable" to false),
            "title" to mapOf("type" to "String", "nullable" to false),
        ), "Blog post")

        val parsed = Json.parseToJsonElement(Schema.exportTypes(pretty = true)).jsonObject
        val types  = parsed["types"]!!.jsonArray
        val names  = types.map { it.jsonObject["name"]!!.jsonPrimitive.content }

        assertEquals(2, types.size, "two types must be exported")
        assertTrue(names.contains("User"), "User must be present")
        assertTrue(names.contains("Post"), "Post must be present")
    }

    @Test
    fun `exported JSON satisfies the schema json structural contract`() {
        Schema.registerType("Order", mapOf(
            "id"     to mapOf("type" to "ID",     "nullable" to false),
            "amount" to mapOf("type" to "Float",  "nullable" to false),
            "status" to mapOf("type" to "String", "nullable" to true),
        ))

        val parsed = Json.parseToJsonElement(Schema.exportTypes(pretty = true)).jsonObject

        // Top-level shape: only `types`
        assertEquals(setOf("types"), parsed.keys,
            "schema.json for types-only export must contain exactly the `types` key")

        // Each type entry must have at minimum `name` and `fields`
        for (t in parsed["types"]!!.jsonArray) {
            val obj = t.jsonObject
            assertTrue(obj.containsKey("name"),   "every type entry must have `name`")
            assertTrue(obj.containsKey("fields"), "every type entry must have `fields`")
        }
    }
}

File: sdks/community/fraiseql-nodejs/tests/schema-roundtrip.test.ts
/**
 * SDK-3: Schema roundtrip golden test.
 *
 * Exercises the full decorator → JSON export pipeline: register a type
 * with fields (including a scoped field), export to JSON, and verify that
 * the output matches the expected schema.json structure exactly.
 *
 * This is the contract between the SDK and the fraiseql-cli compiler.
 * If the SDK produces malformed JSON the compiler rejects it — but
 * without this test that failure is silent during SDK development.
 */
import { describe, it, expect, beforeEach, afterEach } from '@jest/globals';
import { Schema } from '../src/schema';

describe('Schema Roundtrip — full decorator → export pipeline', () => {
  beforeEach(() => Schema.reset());
  afterEach(() => Schema.reset());

  it('produces the expected schema.json structure for a single type', () => {
    // Register a realistic type with a mix of field types, including a scoped field.
    Schema.registerType('Article', {
      fields: {
        id:    { type: 'ID',     nullable: false },
        title: { type: 'String', nullable: false },
        body:  { type: 'String', nullable: true  },
        email: { type: 'String', nullable: false, scope: 'read:Article.email' },
      },
      description: 'A published article',
    });

    const json = Schema.exportTypes(true);
    const parsed = JSON.parse(json);

    // Output must be a valid JSON object
    expect(typeof parsed).toBe('object');

    // Must contain exactly the `types` key — no compiler-reserved keys
    expect(parsed).toHaveProperty('types');
    expect(parsed).not.toHaveProperty('queries');
    expect(parsed).not.toHaveProperty('mutations');
    expect(parsed).not.toHaveProperty('observers');
    expect(parsed).not.toHaveProperty('security');
    expect(parsed).not.toHaveProperty('federation');

    // Exactly one type was registered
    expect(Array.isArray(parsed.types)).toBe(true);
    expect(parsed.types).toHaveLength(1);

    // Verify Article type structure
    const article = parsed.types[0];
    expect(article.name).toBe('Article');
    expect(article.description).toBe('A published article');

    // All four fields must be present
    const fieldNames: string[] = article.fields.map((f: { name: string }) => f.name);
    expect(fieldNames).toContain('id');
    expect(fieldNames).toContain('title');
    expect(fieldNames).toContain('body');
    expect(fieldNames).toContain('email');

    // The scoped field must carry its scope annotation
    const emailField = article.fields.find((f: { name: string }) => f.name === 'email');
    expect(emailField).toBeDefined();
    expect(emailField.scope).toBe('read:Article.email');
  });

  it('exports multiple types with correct names', () => {
    Schema.registerType('User', {
      fields: {
        id:   { type: 'ID',     nullable: false },
        name: { type: 'String', nullable: false },
      },
      description: 'System user',
    });

    Schema.registerType('Post', {
      fields: {
        id:    { type: 'ID',     nullable: false },
        title: { type: 'String', nullable: false },
      },
      description: 'Blog post',
    });

    const parsed = JSON.parse(Schema.exportTypes(true));
    expect(parsed.types).toHaveLength(2);

    const names: string[] = parsed.types.map((t: { name: string }) => t.name);
    expect(names).toContain('User');
    expect(names).toContain('Post');
  });

  it('exported JSON satisfies the schema.json structural contract', () => {
    Schema.registerType('Order', {
      fields: {
        id:     { type: 'ID',    nullable: false },
        amount: { type: 'Float', nullable: false },
        status: { type: 'String', nullable: true },
      },
    });

    const parsed = JSON.parse(Schema.exportTypes(true));

    // Top-level shape: only `types`
    expect(Object.keys(parsed)).toEqual(['types']);

    // Each type entry must have at minimum `name` and `fields`
    for (const t of parsed.types) {
      expect(typeof t.name).toBe('string');
      expect(Array.isArray(t.fields)).toBe(true);
    }
  });
});
